# Contents

## 01 Importing Libraries

## 02 Importing Data

## 03 Procedures from the exercise

## 04 Task 4.5

## Running df.describe() on df_ords

## Checking for mixed-type data in df_ords dataframe

## Checking for missing values in df_ords dataframe

## Addressing the missing values

## Checking for duplicate values

## Addressing the duplicate values

## Exporting data

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

Importing Data

In [2]:
# Creating a path
path = r'C:\Users\ketki\OneDrive\Desktop\Data Analytics bootcamp tasks\Data Immersion\Achievement 4 Python\Instacart Basket Analysis'

In [42]:
#Importing orders.csv to df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [43]:
#Importing products.csv to df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

In [5]:
# Create a dataframe
df_test = pd.DataFrame()

In [6]:
# Create a mixed type column
df_test['mix'] = ['a','b',1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
df_test['mix'] = df_test['mix'].astype('str')

In [12]:
# Importing prods.csv to df_prods
df_prods = pd.read_csv (r'C:\Users\ketki\OneDrive\Desktop\Data Analytics bootcamp tasks\Data Immersion\Achievement 4 Python\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [13]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
 df_prods.shape

(49693, 5)

In [19]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
df_prods_clean.shape

(49677, 5)

In [22]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
df_prods_clean.shape

(49677, 5)

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

In [28]:
# Export data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [37]:
df_ords.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The maximum value for prices is 99999, this can be an error or this data could mean something else.

In [45]:
#Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Found no mixed data types

In [46]:
# Run a check for missing values in your df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [48]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [39]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Addressing Missing Values


The missing values could be the number of first time customers.

In [62]:
df_ords.shape

(3421083, 8)

In [ ]:
df_ords_clean = df_ords

In [65]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [57]:
df_ords_clean

Unnamed: 0  order_id  user_id eval_set  order_number  \
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
5                 5   3367565        1    prior             6   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         order_day_of_week  order_hour_of_day  days_since_prior_order  
1                        3                  7                    15.0  
2                        3                 12                    21.0  
3                        4                  7                    29.0  
4                        4                 15                    28.0  
5                        2                  7                    19.0  
...                    ...                ...                     ...  
3421078                  5                 18                    29.0  
3421079                  4                 10                    30.0  
3421080                  1                 12                    18.0  
3421081                  1                 12                     7.0  
3421082                  6                 14                    30.0  

[3214874 rows x 8 columns]

In [55]:
#Run a check for duplicate values in your df_ords data.
df_dups_ords = df_ords_clean[df_ords_clean.duplicated()]

In [56]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

no duplicates found

Exporting the cleaned files

In [66]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))

In [67]:
df_dups_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'order_cleaned.csv'))